// testrail-integration/lib/common-helpers/test-rail/testrail.implements.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.INewMilestoneImpl = exports.IMilestoneImpl = exports.IProjectUpdateImpl = exports.IProjectImpl = exports.ITestResultFiltersImpl = exports.ITestStatusImpl = exports.ISuiteImpl = exports.INewSuiteImpl = exports.ITemplateImpl = exports.ITestrailUserImpl = exports.INewSectionImpl = exports.ISectionUpdateImpl = exports.ISectionImpl = exports.ICaseFieldConfigImpl = exports.ICaseTypeImpl = exports.ICaseFieldImpl = exports.ITestResultImpl = exports.ITestImpl = exports.INewTestRunImpl = exports.ITestRunImpl = exports.ICaseUpdateImpl = exports.ICaseFiltersImpl = exports.ICaseImpl = exports.INewTestResultsImpl = exports.INewTestResultImpl = void 0;
class INewTestResultImpl {
}
exports.INewTestResultImpl = INewTestResultImpl;
class INewTestResultsImpl {
}
exports.INewTestResultsImpl = INewTestResultsImpl;
class ICaseImpl {
}
exports.ICaseImpl = ICaseImpl;
class ICaseFiltersImpl {
}
exports.ICaseFiltersImpl = ICaseFiltersImpl;
class ICaseUpdateImpl {
}
exports.ICaseUpdateImpl = ICaseUpdateImpl;
class ITestRunImpl {
}
exports.ITestRunImpl = ITestRunImpl;
class INewTestRunImpl {
}
exports.INewTestRunImpl = INewTestRunImpl;
class ITestImpl {
}
exports.ITestImpl = ITestImpl;
class ITestResultImpl {
}
exports.ITestResultImpl = ITestResultImpl;
class ICaseFieldImpl {
}
exports.ICaseFieldImpl = ICaseFieldImpl;
class ICaseTypeImpl {
}
exports.ICaseTypeImpl = ICaseTypeImpl;
class ICaseFieldConfigImpl {
}
exports.ICaseFieldConfigImpl = ICaseFieldConfigImpl;
class ISectionImpl {
}
exports.ISectionImpl = ISectionImpl;
class ISectionUpdateImpl {
}
exports.ISectionUpdateImpl = ISectionUpdateImpl;
class INewSectionImpl {
}
exports.INewSectionImpl = INewSectionImpl;
class ITestrailUserImpl {
}
exports.ITestrailUserImpl = ITestrailUserImpl;
class ITemplateImpl {
}
exports.ITemplateImpl = ITemplateImpl;
class INewSuiteImpl {
}
exports.INewSuiteImpl = INewSuiteImpl;
class ISuiteImpl {
}
exports.ISuiteImpl = ISuiteImpl;
class ITestStatusImpl {
}
exports.ITestStatusImpl = ITestStatusImpl;
class ITestResultFiltersImpl {
}
exports.ITestResultFiltersImpl = ITestResultFiltersImpl;
class IProjectImpl {
}
exports.IProjectImpl = IProjectImpl;
class IProjectUpdateImpl {
}
exports.IProjectUpdateImpl = IProjectUpdateImpl;
var SuiteMode;
(function (SuiteMode) {
    SuiteMode[SuiteMode["SINGLE"] = 1] = "SINGLE";
    SuiteMode[SuiteMode["SINGLE_WITH_BASELINES"] = 2] = "SINGLE_WITH_BASELINES";
    SuiteMode[SuiteMode["MULTIPLE_SUITES"] = 3] = "MULTIPLE_SUITES";
})(SuiteMode || (SuiteMode = {}));
class IMilestoneImpl {
}
exports.IMilestoneImpl = IMilestoneImpl;
class INewMilestoneImpl {
}
exports.INewMilestoneImpl = INewMilestoneImpl;
